File: src/fr/geobert/radis/ui/editor/OperationEditor.java
package fr.geobert.radis.ui.editor;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.util.Log;
import com.actionbarsherlock.view.Menu;
import com.actionbarsherlock.view.MenuInflater;
import com.actionbarsherlock.view.MenuItem;
import fr.geobert.radis.BaseActivity;
import fr.geobert.radis.R;
import fr.geobert.radis.data.Operation;
import fr.geobert.radis.data.ScheduledOperation;
import fr.geobert.radis.db.AccountTable;
import fr.geobert.radis.db.DbContentProvider;
import fr.geobert.radis.db.OperationTable;
import fr.geobert.radis.db.ScheduledOperationTable;
import fr.geobert.radis.tools.Tools;

public class OperationEditor extends CommonOpEditor {
    public static final long NO_OPERATION = 0;
    public static final int OPERATION_EDITOR = 2000;
    private static final String TAG = "OperationEditor";
    private static final int GET_OP = 610;
    protected Operation mOriginalOp;
    private OperationEditFragment mEditFragment;

    public static void callMeForResult(final BaseActivity context, final long opId, final long accountId) {
        final Intent intent = new Intent(context, OperationEditor.class);
        intent.putExtra(PARAM_OP_ID, opId);
        intent.putExtra(AccountEditor.PARAM_ACCOUNT_ID, accountId);
        context.startActivityForResult(intent, OPERATION_EDITOR);
    }

    @Override
    protected void setView() {
        setContentView(R.layout.operation_edit);
        mEditFragment = (OperationEditFragment) getSupportFragmentManager().findFragmentById(R.id.main_edit_pane);
    }

    @Override
    protected void fetchOrCreateCurrentOp() {
        if (mRowId > 0) {
            fetchOp(GET_OP);
        } else {
            mCurrentOp = new Operation();
            mCurrentOp.mAccountId = mCurAccountId;
            populateFields();
        }
    }

    @Override
    protected void populateFields() {
        mEditFragment.populateCommonFields(mCurrentOp);
    }

    private void onCancelClicked() {
        setResult(CommonOpEditor.RESULT_CANCELED);
        finish();
    }

    private void onOkClicked() {
        if (mRowId <= 0 || mOriginalOp != null) {
            StringBuilder errMsg = new StringBuilder();
            if (mEditFragment.isFormValid(errMsg)) {
                fillOperationWithInputs(mCurrentOp);
                saveOpAndExit();
            } else {
                Tools.popError(this,
                        errMsg.toString(), null);
            }
        }
    }

    private void setResAndExit() {
        Intent res = new Intent();
        res.putExtra("opId", mRowId);
        res.putExtra("opDate", mCurrentOp.getDate());
        setResult(RESULT_OK, res);
        finish();
    }

    @Override
    protected void saveOpAndExit() {
        Operation op = mCurrentOp;
        Log.d(TAG, "saveOpAndExit, mRowId : " + mRowId);
        if (mRowId <= 0) {
            mRowId = OperationTable.createOp(this, op, op.mAccountId);
            setResAndExit();
        } else {
            if (op.equals(mOriginalOp)) {
                setResAndExit();
            } else {
                if (op.mScheduledId > 0 && !op.equalsButDate(mOriginalOp)) {
                    UpdateScheduledOp.newInstance(mCurrentOp, mPreviousSum, mRowId).show(getSupportFragmentManager(),
                            "dialog");
                } else {
                    OperationTable.updateOp(this, mRowId, op, mOriginalOp);
                    setResAndExit();
                }
            }
        }
    }

    @Override
    protected void fillOperationWithInputs(Operation operation) {
        mEditFragment.fillOperationWithInputs(operation);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getSupportMenuInflater();
        inflater.inflate(R.menu.confirm_cancel_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.cancel:
                onCancelClicked();
                return true;
            case R.id.confirm:
                onOkClicked();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        outState.putParcelable("originalOp", mOriginalOp);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        mOriginalOp = savedInstanceState.getParcelable("originalOp");
        super.onRestoreInstanceState(savedInstanceState);
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle arg1) {
        CursorLoader l = null;
        if (l == null) {
            switch (id) {
                case GET_OP:
                    l = new CursorLoader(this,
                            Uri.parse(DbContentProvider.OPERATION_JOINED_URI + "/"
                                    + mRowId), OperationTable.OP_COLS_QUERY, null,
                            null, null);
                    break;

                default:
                    break;
            }
        }
        return l;
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        hideProgress();
        switch (loader.getId()) {
            case GET_OP:
                data.moveToFirst();
                mCurrentOp = new Operation(data);
                mOriginalOp = new Operation(data);
                populateFields();
                break;
            default:
                break;
        }

    }

    @Override
    public void onLoaderReset(Loader<Cursor> arg0) {

    }

    protected static class UpdateScheduledOp extends DialogFragment {
        public static UpdateScheduledOp newInstance(Operation currentOp, long previousSum, long rowId) {
            UpdateScheduledOp frag = new UpdateScheduledOp();
            Bundle args = new Bundle();
            args.putLong("previousSum", previousSum);
            args.putLong("rowId", rowId);
            args.putParcelable("currentOp", currentOp);
            frag.setArguments(args);
            return frag;
        }

        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            final OperationEditor act = (OperationEditor) getActivity();
            Bundle args = getArguments();
            final Operation currentOp = args.getParcelable("currentOp");
            final long previousSum = args.getLong("previousSum");
            final long rowId = args.getLong("rowId");
            AlertDialog.Builder builder = new AlertDialog.Builder(act);
            builder.setMessage(R.string.ask_update_scheduling)
                    .setCancelable(false)
                    .setPositiveButton(R.string.update,
                            new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog,
                                                    int which) {
                                    final ScheduledOperation op =
                                            new ScheduledOperation(currentOp, currentOp.mAccountId);
                                    if (ScheduledOperationTable.updateScheduledOp(act, currentOp.mScheduledId, op,
                                            true)) {
                                        AccountTable.updateProjection(act, act.mCurAccountId, previousSum - op.mSum,
                                                op.getDate());
                                    }
                                    ScheduledOperationTable.updateAllOccurences(getActivity(), op, previousSum,
                                            currentOp.mScheduledId);
                                    act.setResAndExit();
                                }
                            })
                    .setNeutralButton(R.string.disconnect,
                            new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog,
                                                    int id) {
                                    currentOp.mScheduledId = 0;
                                    OperationTable.updateOp(act, rowId, currentOp, act.mOriginalOp);
                                    act.setResAndExit();
                                }
                            })
                    .setNegativeButton(R.string.cancel,
                            new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog,
                                                    int id) {
                                    dialog.cancel();
                                }
                            });
            return builder.create();
        }
    }
}

File: app/src/main/kotlin/fr/geobert/radis/ui/editor/OperationEditor.kt
package fr.geobert.radis.ui.editor

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.DialogInterface
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.content.CursorLoader
import android.support.v4.content.Loader
import android.util.Log
import android.view.MenuItem
import fr.geobert.radis.BaseActivity
import fr.geobert.radis.R
import fr.geobert.radis.data.Operation
import fr.geobert.radis.data.ScheduledOperation
import fr.geobert.radis.db.AccountTable
import fr.geobert.radis.db.DbContentProvider
import fr.geobert.radis.db.OperationTable
import fr.geobert.radis.db.ScheduledOperationTable
import fr.geobert.radis.tools.Tools

public class OperationEditor : CommonOpEditor() {
    protected var mOriginalOp: Operation? = null
    private val mEditFragment by lazy { OperationEditFragment() }

    override fun fetchOrCreateCurrentOp(cbk: (Operation) -> Unit) {
        if (mRowId > 0) {
            setTitle(R.string.op_edition)
            onOpFetchedCbks.add(cbk)
            fetchOp(GET_OP)
        } else {
            setTitle(R.string.op_creation)
            val op = Operation()
            op.mAccountId = mCurAccountId
            mCurrentOp = op
            cbk(op)
        }
    }

    override fun inflateFragment() {
        supportFragmentManager.beginTransaction().add(R.id.fragment_cont, mEditFragment).commit()
    }

    //    override fun populateFields() = mEditFragment.populateCommonFields(mCurrentOp as Operation)

    private fun onOkClicked() {
        val op = mCurrentOp
        if ((mRowId <= 0 || mOriginalOp != null) && op != null) {
            val errMsg = StringBuilder()
            if (mEditFragment.isFormValid(errMsg)) {
                fillOperationWithInputs(op)
                saveOpAndExit()
            } else {
                Tools.popError(this, errMsg.toString(), null)
            }
        }
    }

    private fun setResAndExit() {
        val res = Intent()
        val op = mCurrentOp
        if (op != null) {
            mCurrentOp?.mRowId = mRowId
            res.putExtra("operation", op)
            setResult(Activity.RESULT_OK, res)
        } else {
            setResult(Activity.RESULT_CANCELED)
        }
        finish()
    }

    override fun saveOpAndExit() {
        val op = mCurrentOp
        Log.d(TAG, "saveOpAndExit, mRowId : " + mRowId)
        if (op != null) {
            if (mRowId <= 0) {
                mRowId = OperationTable.createOp(this, op, op.mAccountId)
                setResAndExit()
            } else {
                val origOp = mOriginalOp
                if (origOp != null) {
                    if (op.equals(origOp)) {
                        setResAndExit()
                    } else {
                        if (op.mScheduledId > 0 && !op.equalsButDate(origOp)) {
                            UpdateScheduledOp.newInstance(op, mPreviousSum, mRowId).show(supportFragmentManager, "dialog")
                        } else {
                            OperationTable.updateOp(this, mRowId, op, origOp)
                            setResAndExit()
                        }
                    }
                }
            }
        }
    }

    override fun fillOperationWithInputs(operation: Operation) {
        mEditFragment.fillOperationWithInputs(operation)
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.confirm -> {
                onOkClicked()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putParcelable("originalOp", mOriginalOp)
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        mOriginalOp = savedInstanceState.getParcelable<Operation>("originalOp")
        super.onRestoreInstanceState(savedInstanceState)
    }

    override fun onCreateLoader(id: Int, arg1: Bundle): Loader<Cursor> {
        var l: CursorLoader? = null
        if (l == null) {
            when (id) {
                GET_OP -> l = CursorLoader(this, Uri.parse("${DbContentProvider.OPERATION_JOINED_URI}/$mRowId"),
                        OperationTable.OP_COLS_QUERY, null, null, null)

                else -> {
                }
            }
        }
        return l as CursorLoader
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        when (loader.id) {
            GET_OP -> {
                val op = if (data.moveToFirst()) {
                    mOriginalOp = Operation(data)
                    Operation(data)
                } else {
                    val o = Operation()
                    o.mAccountId = mCurAccountId
                    o
                }
                mCurrentOp = op
                onOpFetchedCbks.forEach { it(op) }
            }
            else -> {
            }
        }

    }

    override fun onLoaderReset(arg0: Loader<Cursor>) {
    }

    public class UpdateScheduledOp : DialogFragment() {

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            super.onCreateDialog(savedInstanceState)
            val act = activity as OperationEditor
            val args = arguments
            val currentOp = args.getParcelable<Operation>("currentOp")
            val previousSum = args.getLong("previousSum")
            val rowId = args.getLong("rowId")
            val builder = AlertDialog.Builder(act)
            builder.setMessage(R.string.ask_update_scheduling).setCancelable(false).setPositiveButton(R.string.update, object : DialogInterface.OnClickListener {
                override fun onClick(dialog: DialogInterface, which: Int) {
                    val op = ScheduledOperation(currentOp, currentOp.mAccountId)
                    if (ScheduledOperationTable.updateScheduledOp(act, currentOp.mScheduledId, op, true)) {
                        AccountTable.updateProjection(act, act.mCurAccountId, op.mSum, previousSum, op.getDate(), -1L)
                    }
                    ScheduledOperationTable.updateAllOccurences(activity, op, previousSum, currentOp.mScheduledId)
                    act.setResAndExit()
                }
            }).setNeutralButton(R.string.disconnect, object : DialogInterface.OnClickListener {
                override fun onClick(dialog: DialogInterface, id: Int) {
                    currentOp.mScheduledId = 0
                    OperationTable.updateOp(act, rowId, currentOp, act.mOriginalOp!!)
                    act.setResAndExit()
                }
            }).setNegativeButton(R.string.cancel, object : DialogInterface.OnClickListener {
                override fun onClick(dialog: DialogInterface, id: Int) {
                    dialog.cancel()
                }
            })
            return builder.create()
        }

        companion object {
            public fun newInstance(currentOp: Operation, previousSum: Long, rowId: Long): UpdateScheduledOp {
                val frag = UpdateScheduledOp()
                val args = Bundle()
                args.putLong("previousSum", previousSum)
                args.putLong("rowId", rowId)
                args.putParcelable("currentOp", currentOp)
                frag.arguments = args
                return frag
            }
        }
    }

    companion object {
        public val NO_OPERATION: Long = 0
        public val OPERATION_EDITOR: Int = 2000
        public val OPERATION_CREATOR: Int = 2001
        private val TAG = "OperationEditor"
        private val GET_OP = 610

        public fun callMeForResult(context: BaseActivity, opId: Long, accountId: Long) {
            val intent = Intent(context, OperationEditor::class.java)
            intent.putExtra(CommonOpEditor.PARAM_OP_ID, opId)
            intent.putExtra(AccountEditor.PARAM_ACCOUNT_ID, accountId)
            context.startActivityForResult(intent, if (opId == NO_OPERATION) OPERATION_CREATOR else OPERATION_EDITOR)
        }
    }
}

